Add Foreman inventory

dLobatog opened the pull request:
##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
external inventories

##### ANSIBLE VERSION
```
ansible 2.2.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

##### SUMMARY
Adds an inventory to the list of external inventories in contrib. This inventory allows Ansible users to query information from Foreman and Satellite 6 (https://github.com/theforeman/foreman_ansible_inventory).

Should I add the README here too?

dLobatog committed:
Add Foreman inventory

This commit adds the foreman inventory based on
https://github.com/theforeman/foreman_ansible_inventory
and its configuration file.

chrismeyersfsu commented:
@dLobatog please sprinkle the docs https://github.com/theforeman/foreman_ansible_inventory into the .ini file.

abadger commented on contrib/inventory/foreman.py:
```diff
@@ -0,0 +1,391 @@
+        self.facts = dict()   # Facts of each host
+        self.hostgroups = dict()  # host groups
+
+    def run(self):
```
Typically we organize code so that callers are closer to the bottom and callees are near the top (```__init__``` being the exception... it's always at the top).  This organization puts methods like ```run()``` near the bottom of the class and helpers above it.

abadger commented on contrib/inventory/foreman.py:
```diff
@@ -0,0 +1,391 @@
+import os
+import re
+import requests
+from requests.auth import HTTPBasicAuth
```
Note, we try to organize imports into three blocks (all optional).  stdlib imports at the top, then third party library imports (requests in this case), and finally any imports from ansible itself.

abadger commented on contrib/inventory/foreman.py:
```diff
@@ -0,0 +1,391 @@
+        cache.write(json_data)
+        cache.close()
+
+    @staticmethod
```
Helper functions like this (Ones that don't use self.  Ones that could be marked @staticmethod if they were part of the class). should typically be placed at the toplevel.  So far I see to_safe, json_format_dict, and push.

abadger commented on contrib/inventory/foreman.py:
```diff
@@ -0,0 +1,391 @@
+try:
+    import json
+except ImportError:
+    import simplejson as json
```
The json library entered the python stdlib with python-2.6.  Code that runs on the controller (which includes dynamic inventory scripts) need python-2.6 or greater.  So there's no need to fallback to simplejson here.  the json stdlib library should always be available.

abadger commented on contrib/inventory/foreman.py:
```diff
@@ -0,0 +1,391 @@
+try:
+    import ConfigParser
+except ImportError:
+    import configparser as ConfigParser
```
Comment to explain python2 vs python3 would be good here.

abadger commented on contrib/inventory/foreman.py:
```diff
@@ -0,0 +1,391 @@
+
+
+class ForemanInventory(object):
+    config_paths = [
```
This doesn't really matter since only a single instance of ForemanInventory is created in this script but config_paths should really be an instance attribute (created and given an initial value in ```__init__```) rather than a class attribute.  Class attributes exist a single time per class.  All instances of the class would share the same instance of that attribute.  Instance attributes exist, one per instance and thus can change independently.  Modifying config_paths later in the code is a tip-off that this should be an instance-attribute rather than a class-attribute.

Also, you can probably move the setting from an environment variable to be with this code when you do that.  It would seem to make sense to keep that all together.

bcoca commented on contrib/inventory/foreman.py:
```diff
@@ -0,0 +1,391 @@
+import copy
+import os
+import re
+import requests
```
it would be better to avoid the requests lib, but if you cannot you should at least specify a min version.

abadger commented:
Made some easy to fix style comments.  The main logic looks like it will do what it's supposed to.  It looks like there could be some larger changes that could be made (for instance, I think that --host only affects output here... It does not affect what has to be fetched from foreman).  As long as you're continuing to maintain it, I don't think that's a blocker for getting this merged... more of a note about what could be fixed in the future.

abadger commented:
needs_revision

dLobatog commented:
@abadger Thanks for taking a look - not a Python expert myself, but I think I've addressed most of the comments. The one I wasn't so sure about is @bcoca 's ask to pin requests to a version. I think 1.1 (what CentOS 6 has) is probably safe, but I'm not sure if you wanted something different.

Hope this works - I'll see if I can remove requests from here at some point even, it was just.. convenient at the time. Thanks @chrismeyersfsu @abadger @bcoca for taking a look

dLobatog commented:
BTW CI is picking on a problem on ./lib/ansible/modules/network/ldap_entry.py , not foreman.py

abadger commented on contrib/inventory/foreman.py:
```diff
@@ -0,0 +1,394 @@
+        regex = "[^A-Za-z0-9\_]"
+        return re.sub(regex, "_", word.replace(" ", ""))
+
+    def json_format_dict(self, data, pretty=False):
```
This should be a @staticmethod and self removed.

abadger replied:
Actually, better would be to move this to the toplevel as a regular function

abadger commented on contrib/inventory/foreman.py:
```diff
@@ -0,0 +1,394 @@
+        else:
+            return json.dumps(data)
+
+    def push(self, d, k, v):
```
This should be an @staticmethod nad self removed.

abadger replied:
Same here, make this a regular function

abadger replied:
Actually... Even better: use a defaultdict instead of this.  This is how defaultdicts work:
``` python
from collections import defaultdict

foo = defaultdict(list)
foo[k].append(v)
```

abadger commented on contrib/inventory/foreman.py:
```diff
@@ -0,0 +1,394 @@
+        else:
+            d[k] = [v]
+
+    def run(self):
```
This should probably be the last method in the class.

abadger commented on contrib/inventory/foreman.py:
```diff
@@ -0,0 +1,394 @@
+            self.config_paths.append(os.path.expanduser(os.path.expandvars(env_value)))
+
+    @staticmethod
+    def to_safe(word):
```
Make this a regular function

abadger commented on contrib/inventory/foreman.py:
```diff
@@ -0,0 +1,394 @@
+        self._print_data()
+        return True
+
+    def _read_settings(self):
```
This method doesn't seem to do much.  I would merge it into the run() method.

* read_settings and parse_cli_args don't seem related.  They don't operate on the same instance attributes, for instance.
* It's a bit funny to have a private method (_read_settings()) calling public methods (read_settings() and parse_cli_args()) that aren't usable by anything else.
* If this is merged into run(), it will only add one more line there and will remove 5 lines overall.

abadger commented on contrib/inventory/foreman.py:
```diff
@@ -0,0 +1,394 @@
+        except (ConfigParser.NoOptionError, ConfigParser.NoSectionError):
+            group_patterns = "[]"
+
+        self.group_patterns = eval(group_patterns)
```
I didn't notice this before but you have eval here... What is that for?  It's rather dangerous as it means if someone can get code into the config file for the foreman dynamic inventory, they can then execute whatever code they want as the user running ansible.  We should get rid of this if at all possible.

abadger commented on contrib/inventory/foreman.py:
```diff
@@ -0,0 +1,394 @@
+        self.parse_cli_args()
+        return True
+
+    def _get_inventory(self):
```
No need for this to be a private method.  rename to get_inventory().

abadger commented on contrib/inventory/foreman.py:
```diff
@@ -0,0 +1,394 @@
+    def _get_hosts(self):
+        return self._get_json("%s/api/v2/hosts" % self.foreman_url)
+
+    def _get_hostgroup_by_id(self, hid):
```
This doesn't seem to be called from anywhere.  Should it be deleted?

dLobatog replied:
Yes, it was used to resolve Foreman host parameters on https://github.com/theforeman/foreman_ansible_inventory/commit/9972f3fff005a065ea6ebaee73e9fc66d0cf8961#diff-2b3c7ba3ec283a3b380cff70f61e0290R144 , not needed anymore

abadger commented on contrib/inventory/foreman.py:
```diff
@@ -0,0 +1,394 @@
+        if self.args.refresh_cache:
+            self.update_cache()
+        elif not self.is_cache_valid():
+            self.update_cache()
```
Probably better to write this as
``` python
if self.args.refresh_cache or not self.is_cache_valid():
    self.update_cache()
```

abadger commented:
This is probably what the order of methods should be inside of the class:
* __init__()
* read_settings()
* parse_cli_args()
* _get_json()
* _get_hosts()
* _get_all_params_by_id()
* _resolve_params()
* _get_facts_by_id()
* _get_facts()
* write_to_cache()
* _write_cache()
* update_cache()
* is_cache_valid()
* load_inventory_from_cache()
* load_params_from_cache()
* load_facts_from_cache()
* load_cache_from_cache()
* _get_inventory()
* get_host_info()
* _print_data()
* run()


This puts the entrypoint at the bottom and then everything that helps it to run is above it.

abadger commented:
Okay, I think I've gone over this sufficiently that this is everything that needs to be changed.

dLobatog commented:
@abadger Thanks so much for the detailed review! I think I've gone through each of the things you've commented on , with the exception of https://github.com/ansible/ansible/pull/19510#discussion_r94701824 - I didn't know what you referred to, it looks like a regular function already? 

The latest push contains the updates to your comments, including order of methods, and the improvements @wenottingham made here https://github.com/theforeman/foreman_ansible_inventory/commit/7a55c62a6bc6254e7414ece39da550c9c87c5b91

abadger commented:
@dLobatog Right now it's a method.  My comment is to make it a function instead.

dLobatog commented:
@abadger Updated - the `json_format_dict` function is now out of the `ForemanInventory` class. I wasn't sure where to put it so I put it on top. Thanks for the review!

abadger commented:
Merged to devel for inclusion in 2.3.0.  Thaks @dLobatog !

dLobatog commented:
Thanks @abadger for the review!
